### PR TITLE
Added empty check for email, phone and external id

### PR DESCRIPTION
**Pull request by ajaysubra**
# Description

Have empty identifiers primarily email was causing a lot of 400s on our APIs as well as consuming unnecessary user bandwidth. We were vary of adding any validation in our SDK as it's mainly a pass through but felt this would be a good fix to the problem described above.

# Check List

- [ ] Are you changing anything with the public API?
- [ ] Have you tested this change on real device?
- [ ] Are your changes backwards compatible with previous SDK Versions?
- [ ] Have you added unit test coverage for your changes?
- [ ] Have you verified that your changes are compatible with all the operating system version this SDK currently supports?

# Manual Test Plan

1.

# Supporting Materials

### CODE REVIEW

File: Sources/KlaviyoSwift/KlaviyoState.swift [hunk 1]
```diff
@@ -190,7 +193,7 @@ struct KlaviyoState: Equatable, Codable {
         }
         var attributes = profile.data.attributes
         var location = profile.data.attributes.location ?? .init()
-        var properties = profile.data.attributes.properties.value as? [String: Any] ?? [:]
+        let properties = profile.data.attributes.properties.value as? [String: Any] ?? [:]
```
Comment (ajaysubra) [inline Sources/KlaviyoSwift/KlaviyoState.swift:196]: This is unrelated. I messed this up in the previous release so fixing it now.

Comment (ndurell) [reply Sources/KlaviyoSwift/KlaviyoState.swift:196]: nice - I fixed this in a pr too.

**Review by ndurell — APPROVED** [review]
great tests

**Review by kennyklaviyo — APPROVED** [review]

File: Sources/KlaviyoSwift/KlaviyoState.swift [hunk 2]
```diff
@@ -182,9 +182,17 @@ struct KlaviyoState: Equatable, Codable {
     }
 
     mutating func updateStateWithProfile(profile: Profile) {
-        email = profile.email ?? email
-        phoneNumber = profile.phoneNumber ?? phoneNumber
-        externalId = profile.externalId ?? externalId
+        if let profileEmail = profile.email, !profileEmail.isEmpty, profileEmail != email {
```
Comment (kennyklaviyo) [inline Sources/KlaviyoSwift/KlaviyoState.swift:185]: does this allow setting the email to nil?

**Review by evan-masseau — COMMENTED** [review]
LGTM only nit is to log a warning wherever we're rejecting a value per [slack chat](https://klaviyo.slack.com/archives/C03FDDX9BU1/p1712072128112739)

**Comment by ajaysubra** [comment]
> LGTM only nit is to log a warning wherever we're rejecting a value per [slack chat](https://klaviyo.slack.com/archives/C03FDDX9BU1/p1712072128112739)

Done - 
image

**Comment by ajaysubra** [comment]
> > LGTM only nit is to log a warning wherever we're rejecting a value per [slack chat](https://klaviyo.slack.com/archives/C03FDDX9BU1/p1712072128112739)
> 
> Done - image

I'm testing this. Might not need some of the warnings on the profile setter

File: Sources/KlaviyoSwift/KlaviyoState.swift [hunk 3]
```diff
@@ -487,3 +503,14 @@ extension Profile {
         return profile
     }
 }
+
+extension String {
+    fileprivate func isNotEmptyOrSame(as state: String?, identifier: String) -> Bool {
```
Comment (ajaysubra) [inline Sources/KlaviyoSwift/KlaviyoState.swift:508]: Since I was doing this thrice for three of the identifier setters and then again thrice when setting a profile, I instead created an fileprivate extension on string and calling this to keep the comparison logic in just one location. Call site now looks like - `email.isNotEmptyOrSame(as: self.email, identifier: "email")`

**Comment by ajaysubra** [comment]
@evan-masseau @ndurell @kennyklaviyo sorry made another quick refactor since the comparison logic was all over the place. Tested this using the test app and it looks good.

File: Sources/KlaviyoSwift/KlaviyoState.swift (hunk 3 above)
Comment (evan-masseau) [reply Sources/KlaviyoSwift/KlaviyoState.swift:508]: yeah this is nice.

**Review by evan-masseau — APPROVED** [review]
